### ISSUE TITLE
Statically-chained pipeline elements

### ISSUE THREAD

**Issue by nblumhardt** [emit-rs/emit#13]
A bit of a thought-experiment follows.

The crate now supports the creation of an arbitrary processing pipeline, for example:

```
(level) -> (filter) -> (enrich) -> (write) -> (write) -> (send)
```

Each element in the event processing pipeline is invoked by its predecessor as a trait object (`Box<ChainedElement>` at present).

_Some_ virtual dispatch is going to be needed at the start of the pipeline, but it'd be nice to avoid baking this into the internal chaining if we can avoid it.

Imagining that `PipelineBuilder` became `PipelineBuilder<T>` where `T` is the type of the "tail". Ignoring how the tail is initially specified, this could carry the static type information for the successor element right through the chain, which would need to be built in reverse:

```rust
let (p, flush) = PipelineBuilder::new(...)
    // returns PipelineBuilder<FilterElement>
    .pipe(FilterElement::new_exclude("target", "some_module"))
    // returns PipelineBuilder<FixedPropertyEnricher>
    .pipe(FixedPropertyEnricher::new("app", "Some App"))
    // returns PipelineBuilder<Box<ChainedElement>>
    .detach();
```

Some intermediate types/methods would be required to move this from pseudocode to executable, but the type information is definitely there.

Now, since it doesn't matter if we use a lot of virtual dispatch when constructing the pipeline, it _might_ be possible to to this in forward order, by going back to plain `PipelineBuilder` and having each `pipe()` store a boxed, generic object that can be invoked to create the real static chain (in reverse) when `detach()` or `init()` is called.

There could be some blockers I've overlooked in my partial understanding of the interactions between boxed traits and generics in Rust; it'll be interesting to try to create an implementation to flush them out, anyway :-).

**Comment by nblumhardt** [comment]
After tormenting myself with this problem for the best part of a week, I've concluded that while it's easy to build the chain in the reverse direction, it's very-hard-to-impossible to implement a forward-reading builder syntax to do it (at least in the subset of Rust I am familiar with). The few attempts I've made have all landed at a point where something akin to "generic associated types" seems necessary.

The closest thing to a viable forward syntax is nesting the calls; may be an option to integrate someday if it's desirable. It doesn't seem like `PipelineElement`/`ChainedElement` would need to change to support this.

Returning to this in the future with some spare time :-)
